src/static/js/ace2_inner.js

```diff
@@ -4014,55 +4017,36 @@ function Ace2Inner(){
-              // Oh boy the caret is out of the visible area, I need to scroll the browser window to lineNum.
-              if(evt.which == 37 || evt.which == 38){ // If left or up arrow
-                var newY = caretOffsetTop; // That was easy!
-              }
-              if(evt.which == 39 || evt.which == 40){ // if down or right arrow
-                // only move the viewport if we're at the bottom of the viewport, if we hit down any other time the viewport shouldn't change
-                // NOTE: This behavior only fires if Chrome decides to break the page layout after a paste, it's annoying but nothing I can do
-                var selection = getSelection();
-                // top.console.log("line #", rep.selStart[0]); // the line our caret is on
-                // top.console.log("firstvisible", visibleLineRange[0]); // the first visiblel ine
-                // top.console.log("lastVisible", visibleLineRange[1]); // the last visible line
-                // top.console.log(rep.selStart[0], visibleLineRange[1], rep.selStart[0], visibleLineRange[0]);
-                var newY = viewport.top + lineHeight;
-              }
-              if(newY){
-                setScrollY(newY); // set the scrollY offset of the viewport on the document
+        // scroll to viewport when user presses arrow keys and caret is out of the viewport
+        if((evt.which == 37 || evt.which == 38 || evt.which == 39 || evt.which == 40)){
+          // we use arrowKeyWasReleased to avoid triggering the animation when a key is continuously pressed
+          // this makes the scroll smooth
+          if(!continuouslyPressingArrowKey(type)){
+            // We use getSelection() instead of rep to get the caret position. This avoids errors like when
+            // the caret position is not synchronized with the rep. For example, when an user presses arrow
+            // down to scroll the pad without releasing the key. When the key is released the rep is not
+            // synchronized, so we don't get the right node where caret is.
+            var selection = getSelection();
+
+            if(selection){
+              var arrowUp = evt.which === 38;
+
+              // if percentageScrollArrowUp is 0, let the scroll to be handled as default, put the previous
+              // rep line on the top of the viewport
+              var innerHeight = getInnerHeight();
+              if(scroll.arrowUpWasPressedInTheFirstLineOfTheViewport(arrowUp, rep)){
+                var pixelsToScroll = scroll.getPixelsToScrollWhenUserPressesArrowUp(innerHeight);
+
+                // by default, the browser scrolls to the middle of the viewport. To avoid the twist made
+                // when we apply a second scroll, we made it immediately (without animation)
+                scroll.scrollYPageWithoutAnimation(-pixelsToScroll);
+              }else{
+                scroll.scrollNodeVerticallyIntoView(rep, innerHeight);
```

Comment:
Interesting: this entire `if`-`else` block is calling functions from `scroll` module. It looks like something that would make more sense if extracted into a function of it, isn't it?

Reply:
good idea!